content/en/blog/_posts/2025-01-14-devices-failure-handling/index.md

```diff
@@ -0,0 +1,403 @@
+* [Recreate the Device Manager gRPC server if failed · Issue
+ #128167](https://github.com/kubernetes/kubernetes/issues/128167)  
+* [Retry pod admission on device plugin grpc failures · Issue
+  #128043](https://github.com/kubernetes/kubernetes/issues/128043)
+
+Basically, for every interaction of Kubernetes components, we ensure this
+interaction is reliable.
+
+### Roadmap for failure modes: device failed
+
+For the device failures some patterns are already emerging in common scenarios
+that Kubernetes can support. However, the very first step is to make information
+about failed devices available easier. The step 0 here is the work in the KEP:
+Device Health in Pod Status: [Add Resource Health Status to the Pod Status for
+Device Plugin and DRA · Issue #4680 · kubernetes/enhancements ·
+GitHub](https://github.com/kubernetes/enhancements/issues/4680)
+
+Longer term ideas include to be tested:
+
+* Integrate device failures into Pod Failure Policy.  
+* Node-local retry policies, enabling pod failure policies for Pods with
+  restartPolicy=OnFailure and possibly beyond that.  
+* Ability to unschedule pod, including with the restartPolicy=Always, so it can
+  get a new device allocated.  
+* Add device health to the ResourceSlice used to represent devices in DRA,
+  rather than simply withdrawing an unhealthy device from the ResourceSlice.
+
+### Roadmap for failure modes: container code failed
+
+The main improvements to handle container code failures for AI/ML workloads are
+all targeting cheaper error handling and recovery. The cheapness is mostly
+coming from reuse of pre-allocated resources as much as possible. From reusing
+the Pods by restarting containers in-place, to node local restart of containers
+instead of rescheduling whenever possible, to snapshotting support, and
+re-scheduling prioritizing the same node to save on image pulls.
+
+Consider this scenario: A big training job needs 512 Pods to run. And one of the
+pods failed. It means that all Pods need to be interrupted and synced up to
+restart the failed step. The most efficient way to achieve this generally is to
+reuse as many Pods as possible by restarting them in-place, while replacing the
+failed pod to clear up the error from it. Like demonstrated in this picture:
+
+{{< figure src="inplace-pod-restarts.svg" alt="Restarting one pod while recycling all others in a large training job" >}}
+![](k8s-infra-failures.svg)
+
+It is possible to implement this scenario, but all solutions implementing it are
+fragile due to lack of certain extension points in Kubernetes. Adding these
+extension points to implement this scenario is on the Kubernetes roadmap.
+
+### Roadmap for failure modes: device degradation
+
+There is very little done in this area - there is no clear detection signal,
+very limited troubleshooting tooling, and no built-in semantics to express the
+"degraded" device on Kubernetes. There has been discussion of adding data on
```

Comment:
_not blocking feedback_
I think you're being harsh on Kubernetes here. Nodes can be tainted and vendors defining CRDs can (but don't) define a taint field on the CRD representing their device.

I also don't think a taint such as "this GPU's fan isn't spinning" belongs in core Kubernetes.

Would you be willing to mention taints more here (including the limitations of what we have, such as ResourceSlices not supporting resource taints)?

Reply:
The article as it stands doesn't mention taints at all.

Reply:
@mrunalp wdyt?

Yes, we discuss some health releated markers on resource slice. Maybe it is worth mentioning here